src/TLuaInterpreter.cpp

```diff
@@ -5778,10 +5780,114 @@ int TLuaInterpreter::sendTelnetChannel102(lua_State* L)
 // Documentation: https://wiki.mudlet.org/w/Manual:Lua_Functions#getButtonState
 int TLuaInterpreter::getButtonState(lua_State* L)
 {
-    Host& host = getHostFromLua(L);
-    int state;
-    state = host.mpConsole->getButtonState();
-    lua_pushnumber(L, state);
+    auto& host = getHostFromLua(L);
+    if (!lua_gettop(L)) {
+        // The original function only works in the script for a push-down button
+        // and takes no arguments so provide the backwards compatible behaviour
+        // if that is the case:
+        lua_pushnumber(L, host.mpConsole->getButtonState());
+        return 1;
+    }
+
+    // Otherwise consider the type of the first argument
+    auto argType = lua_type(L, 1);
+    TAction* pItem = nullptr;
+    if (argType == LUA_TNUMBER) {
+        int id = lua_tonumber(L, 1);
+        if (id < 0) {
+            return warnArgumentValue(L, __func__, QStringLiteral("item Id (%1) invalid, it must be equal or greater than zero").arg(id).toUtf8().constData());
+        }
+        pItem = host.getActionUnit()->getAction(id);
+        if (!pItem) {
+            return warnArgumentValue(L, __func__, QStringLiteral("no button item with Id %1 found").arg(id).toUtf8().constData());
+        }
+        if (!pItem->isPushDownButton()) {
+            return warnArgumentValue(L, __func__, QStringLiteral("item Id with %1 is not a push-down button").arg(id).toUtf8().constData());
+        }
+    }
+
+    if (argType == LUA_TSTRING) {
+        QString name = lua_tostring(L, 1);
+        if (name.isEmpty()) {
+            return warnArgumentValue(L, __func__, "item name must not be an empty string");
+        }
+        pItem = host.getActionUnit()->findAction(name);
+        if (!pItem) {
+            return warnArgumentValue(L, __func__, QStringLiteral("no button item with name '%1' found").arg(name).toUtf8().constData());
+        }
+        if (!pItem->isPushDownButton()) {
+            return warnArgumentValue(L, __func__, QStringLiteral("item with name '%1' is not a push-down button").arg(name).toUtf8().constData());
+        }
+    }
```

Comment:
This whole part seems to be pretty much the same as in the set function. I would suggest putting this into a function that's used by both.

Reply:
TBH if the code was to be used in *more* than these *two* cases I'd go along with that idea but the additional complexity of a further function call for just this pair means I think it is a borderline decision and I'd prefer not to do that currently...

Reply:
In the medium term I'd like to have a set of *introspection* (?) functions that can *get*, and for some *set*, the common elements of all the mudlet item types (alias, button/menu/toolbars, keybindings, scripts, timers, triggers) i.e. "name", "command", "ID", "active" {based on both it's own *and* its ancestors' state} and possibly some further details... as such it may then be useful to factor out item identification based on EITHER a numeric ID OR a string name basis. At that point this would be useful to revisit.

Reply:
To me, this is not a borderline decision. It has strings and quite a bit of logic that should stay the same. And the easiest way for that is adding a function.